### ISSUE TITLE
Feature addition: Recommendation and notes

### ISSUE THREAD

**Issue by DUFCLiam** [TotalFreedom/TotalFreedomMod#1507]
We can create a new thing called notes:

Notes system:
***
        Add a note to a player: `/notes add <username> <note>`*

        Removing a note: `To be done via console in the notes.yml file`*
***

Recommendation system:
***
 Recommending someone: `/recommend <username>`**

 Removing a recommendation: `/recommend remove <username>`*

 Checking a players recommendations: `/recommend check <username>`
       
Removing a players entry [To be used for acceptations as they cannot re-use the      recommendations for other ranks](Seniors +): `/recommend entryremove <username>`
***

* = Followed by a confirmed command check (#1428)
** = Can only have one entry on one user and will ask for confirmation

**Why should we add this?:**
Why not xD Jokes aside I think we should add this because we need to know people are not faking recommendations right? So we can use this

**Comment by rovertdude** [comment]
This seems quite good and /notes could be used for better ways to find repeat offenders.

If possible, I have a suggestion for it:
When someone is /gtfo'd or /noob'd or anything, it will add a note to them saying their punishment and who banned them.

**Comment by DUFCLiam** [comment]
Yes, It would auto add it saying `<username> was banned on <date> at <time> for: <reason> by <admins name>`

**Comment by Wild1145** [comment]
The only issue you have is people could easily abuse it, i could log in as someone else and get the notes added to their name, all it would take is someone logging in as someone else, being a pain in the backside, getting all the rec's removed and boom.

**Comment by DUFCLiam** [comment]
If they do they will be banned for imposing as another admin. and they can only use it from the same account it will save the senders IP at the end like this [0.0.0.0] so the admins name may not be the same  but will be IP logged.

**Comment by xItsNickHD** [comment]
@dufcliam, then we'll be giving to much information. Could we do it like [0.0.*.*]?

**Comment by DUFCLiam** [comment]
No, @xItsNickHD  OPs cannot view it, Super  Admin+ can also Seniors get there IP logged as [0.0.0.0] for anonymity.

**Comment by Wild1145** [comment]
@DUFCLiam You are missing my point, I am not saying that the admin would login as someone else, but the player would.

**Comment by DUFCLiam** [comment]
The  player cannot do notes or that admins+ can. @Wild1145

**Comment by DUFCLiam** [comment]
@Wild1145 We are not permitting OPs/IMP to use this command. It will be fully verified admins that have permission for this.

**Comment by Wild1145** [comment]
Yes, so what if I want to leave a note on a player? Say I want to leave a note on the player bob, then someone else logs in as bob, abuses, causes a scene and admins are against them getting admin, they leave a note on their name against it, even though they are not the same player.

**Comment by DUFCLiam** [comment]
Well, that can happen with anyone really, We just need to stop all cases of impersonation then is what you are saying.

Also if they abuse they will request an unban... Then we know its them. and we have there IP.

**Comment by LegendIsAwesomes** [comment]
Maybe store them in a configuration file and make the file  clearable with a command.

**Comment by DUFCLiam** [comment]
Yeah, Seniors + Can do it I forgot we can just clear it all. Maybe number them so somebody can remove it example:

[<TIME>] DUFC_Liam: /notes add markbyron The owner of TotalFreedom.
[<TIME>] DUFC_Liam: /notes check markbyron
[<TIME>] CONSOLE: #1: The owner of TotalFreedom [0.0.0.0]
[<TIME>] markbyron: Remove that now!
[<TIME>] DUFC_Liam: Okay.
[<TIME>] DUFC_Liam: /notes remove markbyron 1

**Comment by Wild1145** [comment]
I think this would work for those that are already admin, however I dont think it would work for OP's, just down to the fact its an offline mode server most of the time, and there is no type of verification for them at all...

**Comment by DUFCLiam** [comment]
Oh yeah, We can make it IP bound.

**Comment by Wild1145** [comment]
Thats great, until you do have someone with a non-static IP (Like myself), if I were just an OP this system would just be a huge fail for me. 

FYI: I am not trying to just be an ass, just make sure this isnt attempted half assed.

**Comment by DUFCLiam** [comment]
Hmm... Got any ideas on how to fix this??

**Comment by DUFCLiam** [comment]
I have a dynamic IP too.

**Comment by DUFCLiam** [comment]
Maybe make a YAML file and save it with there IP, Name and maybe a Verification code?

**Comment by DUFCLiam** [comment]
They enter the password and the notes bind to that name

**Comment by ShadowArmadyl** [comment]
I like this idea but it might be easier if you had the notes for a player listed onto an IP.
For example: /notes check <username>

shadowarmadyl:
- 123.123.12.34
  Is a very loyal player, administrator potential.
- 321.321.32.10
  Was caught repeat griefing, spams advertisements.

**Comment by JeromSar** [comment]
I like this idea.

I think we should assume players don't impose other players and attach this to their username. As the /note utility is an admin only tool, I don't think players would intentionally log in as other to destroy its purpose. Additionally, TFM will have a name-lock feature in the future, which will hopefully prevent it, should it ever happen. These notes can be stored in the player data, of course.

**Comment by Wild1145** [comment]
I dont think we can assume that they will not impose unless the server is in onlinemode... Plus we then come back to the other problem of what happens when they change their name?

**Comment by Adaptified** [comment]
Everyone claims that since we are a public repo that we have to keep every server file public. A password system would work if we didn't put the server's password storage file on the repo.

**Comment by DUFCLiam** [comment]
@JeromSar, Do you want me to bind them to the IP and name or the IP? or Name?

**Comment by JeromSar** [comment]
@DUFCLiam Is this something you would like to implement?

I think it should be attached as a `List<String>` in the `PlayerData` class, as that is already name-bound.

**Comment by DUFCLiam** [comment]
Yes, I will _try_ to implement this. @JeromSar

**Comment by JeromSar** [comment]
@DUFCLiam Actually, it might be necessary to store the name of the admin storing the log. Thus, a list of a new class might be necessary. Please think about the implementation beforehand. :)

**Comment by DUFCLiam** [comment]
I think a new YAML file.

**Comment by JeromSar** [comment]
@DUFCLiam No, this type of information is exactly what PlayerData is for. :)

**Comment by DUFCLiam** [comment]
Ahh, Okay ;) @JeromSar

**Comment by DUFCLiam** [comment]
[NOTE]: I am no longer doing this project; Someone else may handle it.

**Comment by DUFCLiam** [comment]
Bump

**Comment by DUFCLiam** [comment]
Also @Wild1145 we can make it a dynamic IP thing just like old impostor system.

**Comment by Pramire** [comment]
Bump @JeromSar, I think there still is a need for this

**Comment by Wild1145** [comment]
@Pramire It is an accepted issue and has been assigned a milestone. It will be implemented in the appropriate version as per the milestone. Please do not bump such issues in the future as it will only cause us as a moderation / development team more work to review.
